ricecooker/utils/add_file.py

```diff
@@ -0,0 +1,204 @@
+                for chunk in response.iter_content( chunk_size = 1024 ):
+                    if chunk: # filter out keep-alive new chunks
+                        f.write(chunk)
+        except:  # Explicitly, we also want to catch CTRL-C here.
+            print("Catching & deleting bad zip created by quitting")
+            try:
+                os.remove(filename)
+            except FileNotFoundError:
+                pass
+            raise
+
+    else:
+        print ("Already exists in cache")
+    # get the bit before the ; in the content-type, if there is one
+    content_type = response.headers.get('Content-Type', "").split(";")[0].strip()
+    return filename, content_type
+
+def create_node(file_class=None, url=None, filename=None, title=None, license=None, copyright_holder=None, description=""):
+    """
+    Create a content node from either a URL or filename.
+    Which content node is determined by:
+    * the 'file_class' explicitly passed (e.g. VideoFile)
+    * guessing from downloaded mimetype, file extension or magic bytes
+    (see guess_type function)
+
+    Use metadata to automatically fille in licence and copyright_holder details --
+    if they're not provided correctly, things will break downstream
+    """
+
+    mime = None
+    if filename is None:
+        assert url, "Neither URL nor filename provided to create_node"
+        filename, mime = download_file(url)
+
+    if file_class is None:
+        with open(filename, "rb") as f:
+            magic_bytes = f.read(8)[:8]  # increase if we use python_magic
+        file_class = guess_type(mime_type=mime,
+                                extension=guess_extension(url or filename),
+                                magic=magic_bytes)
+        # there is a reasonable chance that the file isn't actually a suitable filetype
+        # and that guess_type will raise an UnidentifiedFileType error.
+    assert file_class
+    print (file_class)
+
+    # Transcode video if necessary
+    if file_class == TranscodeVideo:
+        file_class = VideoFile
+        filename = transcode_video(filename)
+
+    if file_class == TranscodeAudio:
+        file_class = AudioFile
+        filename = transcode_audio(filename)
+
+    # TODO - consider non-MP3 audio files
+
+    # Ensure file has correct extension for the type of file we think it is:
+    # this is a requirement from sushichef.
+    extensions = {VideoFile: ".mp4",
+                  AudioFile: ".mp3",
+                  DocumentFile: ".pdf",
+                  HTMLZipFile: ".zip",}
+    extension = extensions[file_class]
+    if not filename.endswith(extension):
+        new_filename = filename + extension
+        os.rename(filename, new_filename)
+        filename = new_filename
+
+    # print (filename, os.path.getsize(filename))
+
+    # Do not permit zero-byte files
+    assert(os.path.getsize(filename))
+
+    kwargs = {VideoFile: {"ffmpeg_settings": {"max_width": 480, "crf": 28}},
```

Comment:
In the case of TranscodedVideo, won't this cause the video to be converted a second time? Maybe we could integrate the conversion / transcode features into pressurecooker? e.g. tweak the functions here and maybe add an audio-only mode? https://github.com/learningequality/pressurecooker/blob/master/pressurecooker/videos.py

Reply:
I don't know enough about the internals of the ricecooker process to reimplement this -- there were a number of issues such as requirements that files needed to end with particular extensions.

We'd need to ensure that the conversion happened exactly once, even if we're not routinely converting all the videos that are already in the correct format.

As it stands it's a small amount of complexity that I feel could be ignored.